File: android/src/main/java/expo/modules/mlkitocr/ExpoMlkitOcrModule.kt
package expo.modules.mlkitocr

import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoMlkitOcrModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoMlkitOcr")

    AsyncFunction("recognizeText") { uri: String, promise: Promise ->
      val context = appContext.reactContext
        ?: return@AsyncFunction promise.reject("NO_CONTEXT", "React context unavailable", null)

      try {
        val inputImage = InputImage.fromFilePath(context, Uri.parse(uri))
        val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)

        recognizer.process(inputImage)
          .addOnSuccessListener { visionText ->
            val blocks = visionText.textBlocks.map { block ->
              mapOf(
                "text" to block.text,
                "boundingBox" to (block.boundingBox?.let {
                  mapOf("x" to it.left, "y" to it.top, "width" to it.width(), "height" to it.height())
                } ?: emptyMap<String, Int>()),
                "lines" to block.lines.map { line ->
                  mapOf(
                    "text" to line.text,
                    "boundingBox" to (line.boundingBox?.let {
                      mapOf("x" to it.left, "y" to it.top, "width" to it.width(), "height" to it.height())
                    } ?: emptyMap<String, Int>()),
                    "elements" to line.elements.map { element ->
                      mapOf(
                        "text" to element.text,
                        "boundingBox" to (element.boundingBox?.let {
                          mapOf("x" to it.left, "y" to it.top, "width" to it.width(), "height" to it.height())
                        } ?: emptyMap<String, Int>()),
                      )
                    }
                  )
                }
              )
            }
            promise.resolve(mapOf("text" to visionText.text, "blocks" to blocks))
          }
          .addOnFailureListener { e ->
            promise.reject("RECOGNITION_FAILED", e.localizedMessage ?: "Text recognition failed", e)
          }
      } catch (e: Exception) {
        promise.reject("IMAGE_LOAD_FAILED", e.localizedMessage ?: "Failed to load image", e)
      }
    }
  }
}

File: ios/ExpoMlkitOcrModule.swift
import ExpoModulesCore
import MLKitTextRecognition
import MLKitVision

public class ExpoMlkitOcrModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoMlkitOcr")

    AsyncFunction("recognizeText") { (uri: String, promise: Promise) in
      guard let url = URL(string: uri), url.scheme == "file" else {
        promise.reject("INVALID_URI", "Invalid file URI: \(uri)")
        return
      }

      let filePath = url.path
      guard let image = UIImage(contentsOfFile: filePath) else {
        promise.reject("IMAGE_LOAD_FAILED", "Failed to load image from: \(filePath)")
        return
      }

      var visionImage = VisionImage(image: image)
      visionImage.orientation = image.imageOrientation

      let textRecognizer = TextRecognizer.textRecognizer()
      textRecognizer.process(visionImage) { result, error in
        if let error = error {
          promise.reject("RECOGNITION_FAILED", "Text recognition failed: \(error.localizedDescription)")
          return
        }

        guard let result = result else {
          promise.reject("RECOGNITION_FAILED", "No recognition result returned")
          return
        }

        let fullText = result.text
        var blocks: [[String: Any]] = []

        for block in result.blocks {
          var blockDict: [String: Any] = [
            "text": block.text,
            "boundingBox": [
              "x": block.frame.origin.x,
              "y": block.frame.origin.y,
              "width": block.frame.width,
              "height": block.frame.height
            ]
          ]

          var lines: [[String: Any]] = []
          for line in block.lines {
            var lineDict: [String: Any] = [
              "text": line.text,
              "boundingBox": [
                "x": line.frame.origin.x,
                "y": line.frame.origin.y,
                "width": line.frame.width,
                "height": line.frame.height
              ]
            ]

            var elements: [[String: Any]] = []
            for element in line.elements {
              let elementDict: [String: Any] = [
                "text": element.text,
                "boundingBox": [
                  "x": element.frame.origin.x,
                  "y": element.frame.origin.y,
                  "width": element.frame.width,
                  "height": element.frame.height
                ]
              ]
              elements.append(elementDict)
            }

            lineDict["elements"] = elements
            lines.append(lineDict)
          }

          blockDict["lines"] = lines
          blocks.append(blockDict)
        }

        promise.resolve([
          "text": fullText,
          "blocks": blocks
        ])
      }
    }
  }
}
